### PR TITLE
Testing Micronaut Kafka Listener using Testcontainers

### PR DESCRIPTION
**Pull request by sdelamo**
Port of Spring Boot Guide https://testcontainers.com/guides/testing-spring-boot-kafka-listener-using-testcontainers

I added a Test Resources section to show how you can simplify testing.

To generate guide locally:

./gradlew testingMicronautKafkaListenerUsingTestcontainersBuild

Open HTML locally:

open build/dist/index.html

Code available:

build/code/testing-micronaut-kafka-listener-using-testcontainers

[PDF Version](https://github.com/micronaut-projects/micronaut-guides/files/12619142/Testing.Kafka.Listener.using.Testcontainers.with.the.Micronaut.Framework.pdf)

### CODE REVIEW

**Review by timyates — APPROVED** [review]

File: guides/testing-micronaut-kafka-listener-using-testcontainers/metadata.json [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{
+  "title": "Testing Kafka Listener using Testcontainers with the Micronaut Framework",
+  "intro": "This guide shows compares how to test a Kafka Listener using Testcontainers in a Micronaut Framework application.",
```
Comment (sivaprasadreddy) [inline guides/testing-micronaut-kafka-listener-using-testcontainers/metadata.json]: Remove "compares".

File: guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+The generated application has http://www.awaitility.org[Awaitility] library as test dependency
+which we can use for asserting the expectations of an asynchronous process flow.
+
+common:create-app-features-testcontainers.adoc[]
+
+== Testcontainers Dependencies
+
+The generated application contains the following Testcontainers dependencies:
+
+dependency:testcontainers[groupId=org.testcontainers,scope=test]
```
Comment (sivaprasadreddy) [inline guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc]: You don't need to add "org.testcontainers:testcontainers" dependency explicitly. Adding the specific modules like kafka, MySQL etc is sufficient.

File: guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+
+callout:micronaut-test-transactional-false[1]
+<2> Annotate the class with `@Property` to supply the driver class name configuration to the test.
+<3> We have configured the https://java.testcontainers.org/modules/databases/jdbc/[Testcontainers special JDBC URL] to spin up a MySQL container and configure it as a DataSource with Micronaut application context.
+callout:test-containers-disabled-without-docker[4]
+callout:test-instance-per-class[5]
+callout:test-property-provider[6]
+<7> We have used the Testcontainers JUnit 5 Extension annotations *@Testcontainers* and *@Container* to spin up a Kafka container and registered the *bootstrap-servers* location using *TestPropertyProvider* mechanism.
+<8> We have created a Product record in the database before sending the event.
+<9> During the test, we sent a message to the *product-price-changes* topic using *KafkaTemplate* with productCode as key and *ProductPriceChangedEvent* instance as value.
```
Comment (sivaprasadreddy) [inline guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc]: Update `KafkaTemplate` to `ProductPriceChangesClient`.

File: guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+
+=== Simpler Test with Test Resources
+
+Thanks to Test Resources, we can simplify the test as follows:
+
+test:ProductPriceChangedEventHandlerTest[app=testresources]
+
+callout:micronaut-test-transactional-false[1]
+<2> We have created a Product record in the database before sending the event.
+<3> During the test, we sent a message to the *product-price-changes* topic using *KafkaTemplate* with productCode as key and *ProductPriceChangedEvent* instance as value.
```
Comment (sivaprasadreddy) [inline guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc]: Update `KafkaTemplate` to `ProductPriceChangesClient`.

Comment (sdelamo) [reply guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc]: resolved via https://github.com/micronaut-projects/micronaut-guides/pull/1332/commits/a953961322ba0f99e4c29d78b52456a50b44fef3

File: guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc (hunk 3 above)
Comment (sdelamo) [reply guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc]: resolved via resolved via https://github.com/micronaut-projects/micronaut-guides/pull/1332/commits/a953961322ba0f99e4c29d78b52456a50b44fef3

File: guides/testing-micronaut-kafka-listener-using-testcontainers/metadata.json (hunk 1 above)
Comment (sdelamo) [reply guides/testing-micronaut-kafka-listener-using-testcontainers/metadata.json]: Resolved via https://github.com/micronaut-projects/micronaut-guides/pull/1332/commits/017c410a31d9bdd02a2e46cd59d557f8356de71b

File: guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc (hunk 2 above)
Comment (sdelamo) [reply guides/testing-micronaut-kafka-listener-using-testcontainers/testing-micronaut-kafka-listener-using-testcontainers.adoc]: resolved via https://github.com/micronaut-projects/micronaut-guides/pull/1332/commits/922664aa013ceedd232d57b45935f36b74d410b5

**Comment by sdelamo** [comment]
@sivaprasadreddy feedback addressed

**Review by sivaprasadreddy — APPROVED** [review]

**Pull request merged**